---
title: "Posvirus Host Prediction"
author: "Katie Tseng, Dan Becker, Heather Koehler, Pilar Fernandez, Stephanie Seifert"
output:
  pdf_document:
  latex_engine: xelatex
  toc: yes
html_document:
  fig_height: 6
  fig_width: 6
  highlight: tango
  theme: journal
---

```{r, echo=F, message=F}
knitr::opts_chunk$set(eval=F)
```

# Introduction

The following code reproduces the analysis from:

> etc.

# Set system

```{r echo=F, message=T}

#clean environment
rm(list=ls()) 
graphics.off()

#set working directory
setwd("~/Library/CloudStorage/OneDrive-WashingtonStateUniversity(email.wsu.edu)/Fernandez Lab/Projects (Active)/OPV Host Prediction/GitHub/PoxHost/Tseng2022")

```

# Load required packages

```{r echo=F, message=T}

library(ape)
library(dplyr)
library(nlme)
library(tidyverse)
library(vroom)
library(treespace) 
#note: treespace dependencies, https://stackoverflow.com/a/66127391/2554330
    #1. install XQuartz v2.7.11 (https://www.xquartz.org/releases/XQuartz-2.7.11.html); 
    #2. install package 'rgl'
        #>install.packages ("rgl")
        #>options(rgl.useNULL=TRUE)
        #>.rs.restartR()
        #>library(rgl)

## for modeling
library(gbm);
library(ROCR);
library(vegan);
library(plyr);library(dplyr);
library(mvtnorm)
library(xtable)
library(parallel)

```

# Load interaction data

```{r echo=F, message=T}

## (1) load data
load("Tseng2022.RData")
  #combine (hostTraits): mammal traits from combine database, <https://doi.org/10.1002/ecy.3344> (file: "trait_data_imputed.csv")
  #dryad (phylo tree): mammal phylogeny tree from Dryad, <https://doi.org/10.5061/dryad.tb03d03> (path: "Data_S8_finalFigureFiles">"_DATA")
  #opvgenes (virusTraits):
  #vertlife (taxa): mammal species taxonomy from vertlife, <https://data.vertlife.org/>
  #virion (poxdata): host-virus paired interactions from virion, <https://github.com/viralemergence/virion/tree/main/Virion> 

```

# Obtain host-poxvirus interaction data

```{r echo=F, message=T}

## (1) extract host-OPV interactions detected via PCR or isolation from virion
poxdata <- virion %>%
  filter(VirusGenus == "orthopoxvirus" & (DetectionMethod %in% c("PCR/Sequencing","Isolation/Observation"))) 

## (2) exclude if host or virus is NA and if variola virus
poxdata <- poxdata[!is.na(poxdata$Host),]
poxdata <- poxdata[!is.na(poxdata$Virus),]
poxdata <- poxdata[!(poxdata$Virus=="variola virus"),]

## (3) to dis-aggregate West African from Congo Basin MPXV clades, subset MPXV interactions, export, and merge clade data into poxdata
mpxvdata <- poxdata %>% filter(Virus=="monkeypox virus" & (DetectionMethod %in% c("PCR/Sequencing","Isolation/Observation")))
write.csv(mpxvdata, "")

## (4) separate PCR-positive data
pcr <- subset(poxdata[which(poxdata$DetectionMethod=="PCR/Sequencing"),], select=c("Host","Virus"))
pcr$pcr_studies <- 1
pcr <- aggregate(.~Host+Virus, data=pcr, sum)

## (5) separate isolation-positive data
competence <- subset(poxdata[which(poxdata$DetectionMethod=="Isolation/Observation"),], select=c("Host","Virus"))
competence$competence_studies <- 1
competence <- aggregate(.~Host+Virus, data=competence, sum)

## (6) merge PCR/isolation-positive data; create binary vars
poxdata <- merge(pcr, competence, by=c("Host","Virus"), all=TRUE)
poxdata$pcr=ifelse(is.na(poxdata$pcr_studies),0,1)
poxdata$competence=ifelse(is.na(poxdata$competence_studies),0,1)

## (7) create genus variable
poxdata$Host=(str_replace(poxdata$Host," ","_"))
poxdata$Host=str_to_title(poxdata$Host)
poxdata$gen=sapply(strsplit(as.character(str_to_title(poxdata$Host)),"_"),function(x) x[1])

## (8) clean environment
rm(virion,pcr,competence)

```

# Merge taxonomy data

```{r echo=F, message=T}

## (1) simplify taxa to genus-level
taxa <- vertlife
gtaxa=taxa[!duplicated(taxa$gen),]
gtaxa=gtaxa[c('gen','fam','ord','clade','higher')]

## (2) check for mismatched names, then merge
poxdata$gen[!poxdata$gen %in% taxa$gen]
data=merge(gtaxa,poxdata,by='gen',all.x=TRUE)

## (3) create binary var for sampled vs. unsampled
data$sampled=ifelse(is.na(data$pcr) & is.na(data$competence),0,1)

## (4) create pseudo-absences for viral detection
data$pcr=ifelse(is.na(data$pcr),0,data$pcr)
data$competence=ifelse(is.na(data$competence),0,data$competence)

## (5) keep only genera from orders in which positive associations exist
keep=subset(data, pcr==1 | competence==1)
data$keep=ifelse(data$ord %in% keep$ord,TRUE,FALSE)
data=subset(data,keep==TRUE)
data$keep=NULL

## (6) clean environment
rm(poxdata,taxa,gtaxa,keep)

```

# Merge viral accessory genes

```{r echo=F, message=T}

## (1) clean data
viralTraits <- opvgenes
viralTraits <- head(viralTraits, -2)          

## (2) rename column names
original_cols <- colnames(viralTraits)
colnames(viralTraits) <- paste("ag" ,original_cols,sep="_")
names(viralTraits)[1:2] <- c("virus","OPG")

## (3) to assess variation in viralTraits, create mode function
mode.prop <- function(x) {                
  ux <- unique(x[is.na(x)==FALSE])        # creates array of unique values
  tab <- tabulate(match(na.omit(x), ux))  # creates array of the frequency (number of times) a unique value appears in a column 
  max(tab)/length(x[is.na(x)==FALSE])     # max-frequency / number of elements in each column that are not NA
}

## (4) assess variation across columns (2 indicates columns)
vars=data.frame(apply(viralTraits,2,function(x) mode.prop(x)),
                apply(viralTraits,2,function(x) length(unique(x))))    # number of unique elements in each column
colnames(vars) <- c("comp","uniq")

# ## get names
# vars$variables=rownames(vars)
# names(vars)=c("var","uniq","column")

## simplify
vars <- vars[-c(1,2), ]

## label variables "cut" if homogeneous (100%)
# vars$keep=ifelse(vars$var<1,"keep","cut")
# # vars$keep=ifelse(vars$column%in%c('pcr','competence','antibodies','fam'),'keep',vars$keep) # ensures we keep these columns
# vars=vars[order(vars$keep),]
# 
# ## trim (creates array of column names to cut and removes from df)
# keeps=vars[-which(vars$keep=="cut"),]$column
# 
# ## drop if no variation
# data=data[keeps]
# rm(keeps,vars)
# 
# ## assess missing values
# mval=data.frame(apply(data,2,function(x) length(x[!is.na(x)])/nrow(data))) # proportion of values that are not NA
# 
# ## get names
# mval$variables=rownames(mval)
# names(mval)=c("comp","column")

## visualize distribution of NA
png("/Users/katietseng/Downloads/virus_ag_variation.png", width=4,height=4,units="in",res=600)
ggplot(vars,
       aes(comp))+
  geom_histogram(bins=50)+
  geom_vline(xintercept=0.70,linetype=2,size=0.5)+
  theme_bw()+
  theme(panel.grid.major=element_blank(),panel.grid.minor=element_blank())+
  theme(axis.title.x=element_text(margin=margin(t=10,r=0,b=0,l=0)))+
  theme(axis.title.y=element_text(margin=margin(t=0,r=10,b=0,l=0)))+
  labs(y="frequency",
       x="trait coverage across mammal species (genus)")+
  scale_x_continuous(labels = scales::percent)
dev.off()
# 
# ## round values
# mval$comp=round(mval$comp,2)
# 
# ## label variables "cut" if >30% values are NA
# mval$keep=ifelse(mval$comp>=0.70,"keep","cut")
# table(mval$keep)
# mval=mval[order(mval$keep),]
# 
# ## trim (creates array of column names to cut and removes from df)
# keeps=mval[-which(mval$keep=="cut"),]$column
# 
# ## order
# mval=mval[order(mval$comp),]
# 
# ## drop if not well represented
# data=data[keeps]
# rm(keeps,mval)
```


# Aggregate at genus level

```{r echo=F, message=T}
# pox_data <- virion
# taxa <- vertlife
# host_traits <- combine
# virus_traits <- opvgenes
# tree <- dryad
```

# Combine PubMed citations and evolutionary distinctiveness measure

```{r echo=F, message=T}
```

# Phylogenetic analysis

```{r echo=F, message=T}

# 02_phylofactor.R

```

# Inspect and format data for inclusion in the model

```{r echo=F, message=T}

```

# 

```{r echo=F, message=T}
```

# 

```{r echo=F, message=T}
```

# 

```{r echo=F, message=T}
```

# 