---
title: "Posvirus Host Prediction"
author: "Katie Tseng, Dan Becker, Heather Koehler, Pilar Fernandez, Stephanie Seifert"
output:
  pdf_document:
  latex_engine: xelatex
  toc: yes
html_document:
  fig_height: 6
  fig_width: 6
  highlight: tango
  theme: journal
---

```{r, echo=F, message=F}
knitr::opts_chunk$set(eval=F)
```

# Introduction

The following code reproduces the analysis from:

> etc.

# Set system

```{r echo=F, message=T}

#clean environment
rm(list=ls()) 
graphics.off()

#set working directory
setwd("~/Library/CloudStorage/OneDrive-WashingtonStateUniversity(email.wsu.edu)/Fernandez Lab/Projects (Active)/OPV Host Prediction/GitHub/PoxHost/Tseng2022")

```

# Load required packages

```{r echo=F, message=T}

library(ape)
library(dplyr)
library(nlme)
library(tidyverse)
library(vroom)
library(treespace) 
#note: treespace dependencies, https://stackoverflow.com/a/66127391/2554330
    #1. install XQuartz v2.7.11 (https://www.xquartz.org/releases/XQuartz-2.7.11.html); 
    #2. install package 'rgl'
        #>install.packages ("rgl")
        #>options(rgl.useNULL=TRUE)
        #>.rs.restartR()
        #>library(rgl)

## for modeling
library(gbm);
library(ROCR);
library(vegan);
library(plyr);library(dplyr);
library(mvtnorm)
library(xtable)
library(parallel)
```

# Load, format and merge data

```{r echo=F, message=T}

#### LOAD DATA ####

load("Tseng2022.RData")
#virion: host-virus paired interactions from virion, <https://github.com/viralemergence/virion/tree/main/Virion> 
#vertlife (taxa): mammal species taxonomy from vertlife, <https://data.vertlife.org/>
#combine (hostTraits): mammal traits from combine database, <https://doi.org/10.1002/ecy.3344> (file: "trait_data_imputed.csv")
#opvgenes:
#dryad: mammal phylogeny tree from Dryad, <https://doi.org/10.5061/dryad.tb03d03> (path: "Data_S8_finalFigureFiles">"_DATA")

#### OBTAIN POX-DATA ####

## subset host-orthopoxvirus interactions
poxdata <- virion %>%
  filter(VirusGenus == "orthopoxvirus" & (DetectionMethod %in% c("PCR/Sequencing","Isolation/Observation"))) %>% 
#  dplyr::select(Host,Virus,DetectionMethod) %>%
  unique()

# poxdata <- poxdata %>%
#   dplyr::select(Host,Virus,DetectionMethod) %>%
#   unique() 

## exclude NAs and variola virus
poxdata <- poxdata[!is.na(poxdata$Host),]
poxdata <- poxdata[!is.na(poxdata$Virus),]
poxdata <- poxdata[!(poxdata$Virus=="variola virus"),]
poxdata <- subset(poxdata, select=c("Host","Virus","DetectionMethod"))

## subset detection by PCR and virus isolation
pcr <- poxdata[which(poxdata$DetectionMethod=="PCR/Sequencing"),]
pcr$pcr_studies <- 1
pcr$DetectionMethod=NULL
pcr <- aggregate(.~Host+Virus, data=pcr, sum)

competence <- poxdata[which(poxdata$DetectionMethod=="Isolation/Observation"),]
competence$competence_studies <- 1
competence$DetectionMethod=NULL
competence <- aggregate(.~Host+Virus, data=competence, sum)

## merge PCR and isolation data; simplify and recode as binary
poxdata <- merge(pcr, competence, by=c("Host","Virus"), all=TRUE)
poxdata <- subset(poxdata, select=-c(DetectionMethod.x,DetectionMethod.y))
poxdata$pcr=ifelse(is.na(poxdata$pcr),0,poxdata$pcr)
poxdata$competence=ifelse(is.na(poxdata$competence),0,poxdata$competence)

## create genus variable
poxdata$Host=(str_replace(poxdata$Host," ","_"))
poxdata$Host=str_to_title(poxdata$Host)
poxdata$gen=sapply(strsplit(as.character(str_to_title(poxdata$Host)),"_"),function(x) x[1])

## clean environment
rm(virion,pcr,competence)

#### MERGE WITH MAMMAL TAXONOMY ####

## simplify taxa to genus-level
taxa <- vertlife
gtaxa=taxa[!duplicated(taxa$gen),]
gtaxa=gtaxa[c('gen','fam','ord','clade','higher')]

## check for mismatched names before merging
poxdata$gen[!poxdata$gen %in% taxa$gen]
data=merge(gtaxa,poxdata,by='gen',all.x=TRUE)

## sort by taxonomy level (high to low)
data <- data[order(data$higher,data$clade,data$ord,data$fam,data$gen),]

## note genera that are unsampled/pseudoabsences
data$studied=ifelse(is.na(data$pcr) & is.na(data$competence),0,1)

## create pseudo-absences for viral detection
data$pcr=ifelse(is.na(data$pcr),0,data$pcr)
data$competence=ifelse(is.na(data$competence),0,data$competence)

## subset data of pcr/isolation positive associations only
keep=subset(data, pcr==1 | competence==1)

## create variable denoting whether the order of genera with positive associations exists in merged data 
data$keep=ifelse(data$ord %in% keep$ord,TRUE,FALSE)

## subset true values
data=subset(data,keep==TRUE)

## clean dataset
data=subset(data, select=-c(antibodies,keep))

## clean environment
rm(poxdata,taxa,gtaxa,keep)

```


# Obtain host-poxvirus interaction data

```{r echo=F, message=T}

#subset host-orthopoxvirus interactions
poxdata <- virion %>%
  filter(VirusGenus == "orthopoxvirus" & (DetectionMethod %in% c("PCR/Sequencing","Isolation/Observation","Antibodies"))) %>% 
  dplyr::select(Host,Virus,DetectionMethod) %>%
  unique()

#exclude NAs and variola virus
poxdata <- poxdata[!is.na(poxdata$Host),]
poxdata <- poxdata[!is.na(poxdata$Virus),]
poxdata <- poxdata[!(poxdata$Virus=="variola virus"),]

## subset detection by PCR
pcr <- poxdata[which(poxdata$DetectionMethod=="PCR/Sequencing"),]
pcr$pcr <- 1

## subset detection by virus isolation
competence <- poxdata[which(poxdata$DetectionMethod=="Isolation/Observation"),]
competence$competence <- 1

## merge PCR & isolation data
poxdata <- merge(pcr, competence, by=c("Host","Virus"), all=TRUE)

## simplify dataset
poxdata <- subset(poxdata, select=-c(DetectionMethod.x,DetectionMethod.y))

## recode as binary (replace NA with zeroes)
poxdata$pcr=ifelse(is.na(poxdata$pcr),0,poxdata$pcr)
poxdata$competence=ifelse(is.na(poxdata$competence),0,poxdata$competence)

## create genus variable
poxdata$Host=(str_replace(poxdata$Host," ","_"))
poxdata$Host=str_to_title(poxdata$Host)
poxdata$gen=sapply(strsplit(as.character(str_to_title(poxdata$Host)),"_"),function(x) x[1])

## clean environment
rm(virion,pcr,competence)

```

# M
# Aggregate at genus level

```{r echo=F, message=T}
# pox_data <- virion
# taxa <- vertlife
# host_traits <- combine
# virus_traits <- opvgenes
# tree <- dryad
```

# Combine PubMed citations and evolutionary distinctiveness measure

```{r echo=F, message=T}
```

# Phylogenetic analysis

```{r echo=F, message=T}

# 02_phylofactor.R

```

# Inspect and format data for inclusion in the model

```{r echo=F, message=T}

```

# 

```{r echo=F, message=T}
```

# 

```{r echo=F, message=T}
```

# 

```{r echo=F, message=T}
```

# 